doc: add documentation on the substrafl workflow and compute plan execution (#173)
Substra/substra-documentation: docs/source/documentation/concepts.rst

.. TODO:: Detail vertical and horizontal scaling

.. TODO:: Explain what is substra tools

Compute plan execution
-----------------------

Read this paragraph to understand what happens during the compute plan execution and what can be done to improve the execution time.

Once a compute plan is submitted to the platform, its tasks are scheduled to be executed on each organization.

On each organization, Substra fetches the assets needed for the first task, builds the Docker image of the algo and creates a container with the relevant assets. The task executes and Substra saves its outputs.

> Afterwards, every task **from the same compute plan** that uses the same algo is executed in the same Docker container.

Reviewer: ```suggestion
Afterwards, every task **from the same compute plan** that uses the same Algo is executed in the same Docker container.
```
Maybe capitalising here emphasise the fact that it is a Substra concept.

> Afterwards, every task **from the same compute plan** that uses the same algo is executed in the same Docker container.

Reviewer: ```suggestion
Afterwards, every task **from the same compute plan** that uses the same algo is executed in the same container.
```
I would remove docker since it's a company name and the abstract notion of container should be enough.

> Afterwards, every task **from the same compute plan** that uses the same algo is executed in the same Docker container.

Reviewer 2: This is not a always a docker container:
```suggestion
Afterwards, every task **from the same compute plan** that uses the same algo is executed in the same container.
```

Reviewer 3: What is it then if it's not a Docker container?

Reviewer 2: It can be CRI-O or containerd, kubernetes supports several [container runtimes](https://kubernetes.io/docs/setup/production-environment/container-runtimes/)

Reviewer 4: And it's mainly containerd that we are using !

Reviewer 3: How do we choose between Docker and containerd? Does it depend on the task?

And the user does provide a Dockerfile for each task. So are these dockerfile then "translated" somehow to be used by other container runtimes?

Reviewer 3: I discussed this with @AlexandrePicosson, it's much clearer now.

Asset preparation
^^^^^^^^^^^^^^^^^^

The first step of the task execution is to fetch the necessary assets.

> These include the inputs (e.g. the algo or opener files), the output of other tasks (input artifacts of the task) and data samples.

Reviewer: I don't know if it's just the GitHub rendering but the e.g. part parenthesis and dots are not visible.

Author: they render well on the doc page: https://owkin-substra-documentation--173.com.readthedocs.build/en/173/documentation/concepts.html#compute-plan-execution

The assets, data samples excluded, come from the file systems of the organizations. If they are stored on other organizations, they are downloaded over HTTPS connections.

> Example: an algo submitted on another organization

Reviewer: ```suggestion
Example: an algo submitted on another organization.
```

Example: an algo submitted on another organization.

> The data samples are stored on the organization, in a storage solution (MiniO). They are downloaded for the task, this may take a long time if the dataset is large.

Author: > The data samples are stored **on the organization**

is this enough? How would you say it?

> The data samples are stored on the organization, in a storage solution (MiniO). They are downloaded for the task, this may take a long time if the dataset is large.

Reviewer 3: Ok to say things differently, for a user without knowledge about how substra works, **download** will imply going outside or inside an organisation.

```suggestion
The data samples are stored on the organization, in a storage solution (MiniO). They are downloaded from the organization storage solution (Mini0) to the organization filesystem? task worskspace?), this may take a long time if the dataset is large.
Note that data samples never leave the organization.
```

> The data samples are stored on the organization, in a storage solution (MiniO). They are downloaded for the task, this may take a long time if the dataset is large.

Author: thank you for the suggestion, I changed the sentence, is it better?

> All the organization data is stored on the organization storage solution (MiniO). The task data samples are downloaded to the organization filesystem which may take a long time if the dataset is large.

Reviewer 3: ```suggestion
All the organization data is stored on the organization storage solution (MiniO). The task data samples are downloaded from the organization storage solution to the organization filesystem which may take a long time if the dataset is large.
Note that data samples never leave the organization.
```

> Since this step can be quite long, there is a cache system: on a given organization, all the downloaded files (assets and data samples) are saved on disk. So when another tasks reuses the same assets there is no need to download them again. One the cache is full, all its contents are deleted.

Reviewer: > One the cache is full, all its contents are deleted.

"Once the cache is full, the worker will delete all its content." sounds better to me but feel free to adapt.

Docker image build
^^^^^^^^^^^^^^^^^^^

> For the first task of the compute plan that uses a given algo, Substra needs to build the Docker image and create the

Reviewer: ```suggestion
For the first task of the compute plan that uses a given algo, Substra needs to build the Container image and create the
```

> Docker container. This may take from a few minutes to an hour (or more), depending on the Docker image.

Reviewer: An hour would be quite pessimistic, unless you install a lot of stuff in your container.

Reviewer 4: It also depends on the size of the final image that need to be transfered to the private registry and then to the local kubernetes registry

Author: also I changed the previous sentence to "Substra needs to build the image, transfer it to the local registry then use it to spawn the container." 
I don't want to go into too much detail on the registry, as it seems to me like an implementation detail, and users are not familiar with registries

> For the first task of the compute plan that uses a given algo, Substra needs to build the image, transfer it to the local registry then use it to spawn the container. This takes a few minutes for a small image, it may take a lot of time for larger images.

Author: yes a local Docker registry

> For the first task of the compute plan that uses a given algo, Substra needs to build the image, transfer it to the local registry then use it to spawn the container. This takes a few minutes for a small image, it may take a lot of time for larger images.

Author: Updated the sentence to `local image registry` since following the previous comments I prefer not to say `Docker` more than necessary (and since there are actually two internal registries and I don't want to go into those details)
Do you think it is ok this way, or should we find a way to say there is an additional step here without talking about registries?

> For the first task of the compute plan that uses a given algo, Substra needs to build the image, transfer it to the local registry then use it to spawn the container. This takes a few minutes for a small image, it may take a lot of time for larger images.

Reviewer 3: Good for me like that, Thanks!

.. TODO:: Detail vertical and horizontal scaling

.. TODO:: Explain what is substra tools

Compute plan execution
-----------------------

Read this paragraph to understand what happens during the compute plan execution and understand what can be done to improve the execution time.

> Once a compute plan is submitted to the platform, its tasks are scheduled to be executed on each organisation.

Reviewer: ```suggestion
Once a compute plan is submitted to the platform, its tasks are scheduled to be executed on each organization.
```
I think we use the American version everywhere else.

Author: yes, thanks !
